src/main/java/io/strimzi/kafka/bridge/mqtt/core/MqttKafkaMapper.java

```diff
@@ -0,0 +1,161 @@
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.bridge.mqtt.core;
+
+import io.strimzi.kafka.bridge.mqtt.utils.MappingRule;
+
+import java.util.List;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Comparator;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Responsible for handling all the topic mapping.
+ *
+ * @see MappingRule
+ */
+public class MqttKafkaMapper {
+
+    // default kafka topic. Used when no mapping rule matches the mqtt topic.
+    public static final String DEFAULT_KAFKA_TOPIC = "messages_default";
+
+    // find any word inside a curly bracket. E.g. {something}, this is known as a placeholder.
+    private static final String MQTT_TOPIC_PLACEHOLDER_REGEX = "\\{\\w+\\}";
+
+    // Regex used to replace the placeholders with {something} in the mqtt pattern.
+    private static final String PLACEHOLDER_REGEX = "(.*)";
+
+    // identifies a single level wildcard character in the mqtt pattern. E.g. sensors/+/data
+    private static final String MQTT_TOPIC_SINGLE_LEVEL_WILDCARD_CHARACTER = "+";
+
+    // Regex expression used to replace the + in the mqtt pattern.
+    private static final String SINGLE_LEVEL_WILDCARD_REGEX = "[^/]+";
+
+    // identifies a multi level wildcard character in the mqtt pattern. E.g. sensors/#
+    private static final String MQTT_TOPIC_MULTI_LEVEL_WILDCARD_CHARACTER = "#";
+
+    // matches any character after the string. Used to replace the # in the mqtt pattern.
+    private static final String MULTIPLE_LEVEL_WILDCARD_REGEX = ".*";
+    private final List<MappingRule> rules;
+    private final List<Pattern> patterns = new ArrayList<>();
+
+    /**
+     * Constructor
+     * <p>
+     * Creates a new instance of MqttKafkaMapper.
+     */
+    public MqttKafkaMapper(List<MappingRule> rules) {
+        this.rules = rules;
+        this.rules.sort(Comparator.comparing(MappingRule::getMqttTopicPatternLevels).reversed());
+        buildRegex();
+    }
+
+    /**
+     * Maps an MQTT topic to a Kafka topic. The topic is mapped according to the defined mapping rules.
+     *
+     * @param mqttTopic represents an MQTT topic.
+     * @return a valid Kafka topic.
+     * @see MappingRule
+     */
+    public String map(String mqttTopic) {
+        for (MappingRule rule : this.rules) {
+            Matcher matcher = this.patterns.get(this.rules.indexOf(rule)).matcher(mqttTopic);
+            HashMap<String, String> placeholders = new HashMap<>();
+
+            if (matcher.matches()) {
+                String mappedKafkaTopic = rule.getKafkaTopicTemplate();
+                String[] mqttTopicPatternParts = rule.getMqttTopicPattern().split("/");
+
+                // find MQTT_TOPIC_PLACEHOLDER_REGEX in the kafkaTopicTemplate.
+                Matcher placeholderMatcher = patterns.get(patterns.size() - 1).matcher(rule.getKafkaTopicTemplate());
```

Comment:
Would it be simpler if  we defined another field like this: 
```
private final static Pattern MQTT_TOPIC_PLACEHOLDER_PATTERN = Pattern.compile(MQTT_TOPIC_PLACEHOLDER_REGEX);
```
and then did this:
```suggestion
                Matcher placeholderMatcher = MQTT_TOPIC_PLACEHOLDER_PATTERN.matcher(rule.getKafkaTopicTemplate());
```
Maybe I am just complicating things, let me know what you think!

Reply:
@kyguy I have already made this change.